presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ReorderJoins.java

```diff
@@ -0,0 +1,346 @@
+import java.util.Set;
+import java.util.function.BiFunction;
+import java.util.stream.IntStream;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+
+import static com.facebook.presto.SystemSessionProperties.getJoinDistributionType;
+import static com.facebook.presto.SystemSessionProperties.getJoinReorderingStrategy;
+import static com.facebook.presto.sql.ExpressionUtils.and;
+import static com.facebook.presto.sql.ExpressionUtils.combineConjuncts;
+import static com.facebook.presto.sql.analyzer.FeaturesConfig.JoinReorderingStrategy.COST_BASED;
+import static com.facebook.presto.sql.planner.EqualityInference.createEqualityInference;
+import static com.facebook.presto.sql.planner.iterative.rule.MultiJoinNode.toMultiJoinNode;
+import static com.facebook.presto.sql.planner.plan.Assignments.identity;
+import static com.facebook.presto.sql.planner.plan.JoinNode.DistributionType.PARTITIONED;
+import static com.facebook.presto.sql.planner.plan.JoinNode.DistributionType.REPLICATED;
+import static com.facebook.presto.sql.planner.plan.JoinNode.Type.INNER;
+import static com.facebook.presto.sql.tree.BooleanLiteral.TRUE_LITERAL;
+import static com.facebook.presto.sql.tree.ComparisonExpressionType.EQUAL;
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.base.Predicates.in;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.collect.ImmutableSet.toImmutableSet;
+import static com.google.common.collect.Iterables.getOnlyElement;
+import static java.util.Objects.requireNonNull;
+
+public class ReorderJoins
+        implements Rule
+{
+    private static final Logger log = Logger.get(ReorderJoins.class);
+
+    private final CostComparator costComparator;
+
+    public ReorderJoins(CostComparator costComparator)
+    {
+        this.costComparator = requireNonNull(costComparator, "costComparator is null");
+    }
+
+    @Override
+    public Optional<PlanNode> apply(PlanNode node, Lookup lookup, PlanNodeIdAllocator idAllocator, SymbolAllocator symbolAllocator, Session session)
+    {
+        if (!(node instanceof JoinNode) || getJoinReorderingStrategy(session) != COST_BASED) {
+            return Optional.empty();
+        }
+
+        JoinNode joinNode = (JoinNode) node;
+        // We check that join distribution type is absent because we only want to do this transformation once (reordered joins will have distribution type already set).
+        if (!(joinNode.getType() == INNER) || joinNode.getDistributionType().isPresent()) {
+            return Optional.empty();
+        }
+
+        MultiJoinNode multiJoinNode = toMultiJoinNode(joinNode, lookup);
+        return Optional.of(new JoinEnumerator(idAllocator, symbolAllocator, session, lookup, multiJoinNode.getFilter(), costComparator).chooseJoinOrder(multiJoinNode.getSources(), multiJoinNode.getOutputSymbols()));
```

Comment:
Possible improvement.
If the chosen plan has cross join and original `JoinNode` also has cross join then maybe it's better to choose original since it could be better (tweaked manually) than arbitrary selected cross join in `createCrossJoins`